using Arm.GrcApi.Modules.VulnerabilityManagement;
using Arm.GrcTool.InfrastruCTOre;
using Arm.GrcTool.Infrastructure;
using GrcApi.Modules.Shared.Helpers;
using System.Security.Claims;

namespace Arm.GrcApi.Modules.VulnerabilityManagement
{
    /*
       * Original Author: Sodiq Quadre
       * Date Created: 05/31/2025
       * Development Group: GRCTools
       * Internal Vulnerability reject endpoint
       */
    public class RejectVulnerabilityEndpoint
    {
        /// <summary>
        /// Internal Vulnerability reject endpoint
        /// </summary>
        /// <param name="payload"></param>
        /// <param name="repo"></param>
        /// <param name="EmailService"></param>
        /// <param name="config"></param>
        /// <param name="user"></param>
        /// <param name="currentUserService"></param>
        /// <returns></returns>
        public static async Task<IResult> HandleAsync(
           RejectVulnerabilityReq payload, IRepository<InternalVulnerability> repo, IEmailService EmailService, IConfiguration config, ClaimsPrincipal user, ICurrentUserService currentUserService
        )
        {
            try
            {
                string requesterName = currentUserService.CurrentUserFullName;
                var update = repo.GetContextByConditon(r => r.Id == payload.InternalVulnerabilityId && r.Status == VulnerabilityStatus.Resolved).FirstOrDefault();
                if (update is null) return TypedResults.NotFound("Internal Vulnerability was not found or it has not been resolved");
                if (update.Action == VulnerabilityStatus.Approve)
                {
                    return TypedResults.BadRequest("Internal Vulnerability has been previously approved");
                }
                update.RejectVulnerability(payload.Comment);
                update.SetModifiedBy(requesterName);
                update.SetModifiedOnUtc(DateTime.Now);
                repo.SaveChanges();

                return TypedResults.Ok("Rejected Successfully");
            }
            catch (Exception ex)
            {

                return TypedResults.BadRequest("Unable to rejected the vulnerability");
            }
        }
    }
}
